### PR TITLE
feat(directive): notify directive before they get destroyed

### PR DESCRIPTION
**Pull request by vsavkin**

### CODE REVIEW

**Comment by vicb** [comment]
@vsavkin just curious why you've chosen to declare `lifecycle` in the annotation - instead of implemented a class (which would have an `onDestroy` method) ?

**Comment by vsavkin** [comment]
@vicb an interface in TypeScript (and AtScript) does mean the same thing as in Dart. In AtScript you cannot ask if an object implements an interface. You can only ask if it extends a class. You can only extend only one class, but there are many lifecycle hooks.

**Comment by vicb** [comment]
@vsavkin it seems like in TS you can explicitly implements `interface`s (by default structural type checking is used). May be we should add this feature to AtScript and add the implemented interfaces/classes to the reflection information on the class ?

What about having `lifecycle` being a  `List<LifeCycleEvent>` ? It could be more robust than the current `List<String>` i.e. if you have a typo in the event name.

**Comment by vsavkin** [comment]
@vicb 
>> May be we should add this feature to AtScript and add the implemented interfaces/classes to the reflection information on the class ?

It can cause confusion. Consider the following:

```
interface Person {
  name:string;
}

class Alice {
  name:string;
}

class Bob implements Person {
  name:person;
}


// currently works
var p1:Person = new Alice();
var p2:Person = new Bob();

p1 instanceof Person // false
p2 instanceof Person // false

```

Since we cannot use instanceof, we would have to introduce another way to check if an object implements an interface. Something like:

```
p1.constructor.retainedInterfaces === []
p2.constructor.retainedInterfaces === [Person]
```

Since `retainedInterfaces` is not the same as assignable, it can be confusing, and imo we should not do that. We had several conversations about it, and, as far as I understand, the consensus is not to do it.

>> List<LifeCycleEvent> 

I think this is a good idea.

File: modules/core/src/compiler/element_injector.js [hunk 1]
```diff
@@ -105,6 +107,28 @@ class DirectiveDependency extends Dependency {
...
 
+export class DirectiveBinding extends Binding {
+  onDestroy:boolean;
+
+  constructor(key:Key, factory:Function, dependencies:List, providedAsPromise:boolean, onDestroy:boolean) {
+    super(key, factory, dependencies, providedAsPromise);
+    this.onDestroy = onDestroy;
+  }
+
+  static createFrom(b:Binding, annotation:Directive):Binding {
```
Comment (rkirov) [inline modules/core/src/compiler/element_injector.js]: createFromBinding?

File: modules/core/src/compiler/element_injector.js [hunk 2]
```diff
@@ -271,16 +296,19 @@ export class ElementInjector extends TreeNode {
...
-    this._obj3 = null;
-    this._obj4 = null;
-    this._obj5 = null;
-    this._obj6 = null;
-    this._obj7 = null;
-    this._obj8 = null;
-    this._obj9 = null;
+    var p = this._proto;
+
+    if (isPresent(p._binding0) && p._binding0.onDestroy) {this._obj0.onDestroy(); this._obj0 = null;}
```
Comment (rkirov) [inline modules/core/src/compiler/element_injector.js]: if `p._binding0` is present but `p._binding0.onDestroy = false` we would not set `this._obj0 = null`.

File: modules/core/src/compiler/view.js [hunk 3]
```diff
@@ -172,7 +172,9 @@ export class View {
 
     // elementInjectors
     for (var i = 0; i < this.elementInjectors.length; i++) {
-      this.elementInjectors[i].clearDirectives();
+      if (isPresent(this.elementInjectors[i])) {
```
Comment (rkirov) [inline modules/core/src/compiler/view.js]: Took care of this #416

File: modules/core/test/compiler/element_injector_spec.js [hunk 4]
```diff
@@ -69,6 +70,18 @@ class NeedsView {
   }
 }
 
+class DirectiveWithDestroy {
+  onDestroyCalled:boolean;
+
+  constructor(){
+    this.onDestroyCalled = false;
```
Comment (rkirov) [inline modules/core/test/compiler/element_injector_spec.js]: I recommend a counter, to assert we only call it once.

File: modules/core/test/compiler/element_injector_spec.js [hunk 5]
```diff
@@ -313,13 +330,24 @@ export function main() {
...
           '(A_Needs_B -> B_Needs_A -> A_Needs_B)');
       });
+
+      it("should call onDestroy on directives subscribed to this event", function () {
+        var inj = injector([
+          DirectiveBinding.createFromType(DirectiveWithDestroy, new Directive({lifecycle: ['onDestroy']}))
+        ]);
+        var destroy = inj.get(DirectiveWithDestroy);
+        inj.clearDirectives();
+        expect(destroy.onDestroyCalled).toBeTruthy();
```
Comment (rkirov) [inline modules/core/test/compiler/element_injector_spec.js]: maybe call 'inj.clearDirectives();' one more time and assert the method is not called again.

**Comment by rkirov** [comment]
+1 to using symbol `onDestroy` instead of string `"onDestroy"`. It would also be more minification friendly. Otherwise, LGTM.

**Pull request merged**